### PR TITLE
Fix for A5-1-3 false positives

### PR DESCRIPTION
**Pull request by rak3-sh**
## Description

A5-1-3's original implementation raised alerts for cases where lambdas had an explicit argument list. However, it raised a false positive in case the lambda had an argument list but was defined inside a macro.

```cpp
#define PARAM_MACRO [](int i) { i; };

int test_lambda_in_macro()
{
  PARAM_MACRO // False positive here.
  return 0;
}
```
As a fix, A5-1-3 is modified to consider only those lambdas which have zero arguments, as non-zero arguments in a lambda function would require the user to specify an explicit argument list. Adding this check, removed all of the false positives that were seen in the project for this rule.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - A5-1-3

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/src/rules/A5-1-3/LambdaExpressionWithoutParameterList.ql [hunk 1]
```diff
@@ -21,6 +21,10 @@ where
   not isExcluded(lambda, LambdasPackage::lambdaExpressionWithoutParameterListQuery()) and
   lambdaFunction = lambda.getLambdaFunction() and
   not lambdaFunction.isAffectedByMacro() and
+  // If it has a parameter, then it will have an
+  // explicit parameter list. Therefore, proceed to check only if the lambda
+  // does not have any parameters.
+  not exists (lambdaFunction.getAParameter()) and
```
Comment (lcartey) [inline cpp/autosar/src/rules/A5-1-3/LambdaExpressionWithoutParameterList.ql:27]: Can you please provide an updated test case? For example, add a macro that expands to a lambda with parameters:
```cpp
#define PARAM_MACRO [](int i) { i; };
```
Then expand that in the test function:
```cpp
PARAM_MACRO // COMPLIANT
```

**Review by lcartey — CHANGES_REQUESTED** [review]
Thanks @rak3-sh, this looks like a good improvement. Just requires a small change to the test case, and an update to the change note.

File: cpp/autosar/src/rules/A5-1-3/LambdaExpressionWithoutParameterList.ql (hunk 1 above)
Comment (rak3-sh) [reply cpp/autosar/src/rules/A5-1-3/LambdaExpressionWithoutParameterList.ql:27]: Thanks for the suggestion. I've added this in: [a7dd290](https://github.com/github/codeql-coding-standards/pull/348/commits/a7dd290a650f9997c97763680bbbb8fa3574e0db)

**Review by lcartey — APPROVED** [review]

**Comment by lcartey** [comment]
@rak3-sh you need to format the test file using `clang-format` - you can see the formatting failure here:
https://github.com/github/codeql-coding-standards/actions/runs/6019907111/job/16339180294?pr=348

**Comment by rak3-sh** [comment]
Formatted using clang-format in https://github.com/github/codeql-coding-standards/pull/348/commits/4297f0a82e461f2076f3db0834d8e79eadb7b8b3

**Pull request merged**